### PR TITLE
Use ParserDatabase in lift.rs

### PR DESCRIPTION
**Pull request by tomhoule**
Using already-known information about the schema there simplifies the
lifting process.

The next step will be the attribute validators.

### CODE REVIEW

File: libs/datamodel/core/src/transform/ast_to_dml/lift.rs [hunk 1]
```diff
@@ -212,204 +178,181 @@ impl<'a> LiftAstToDml<'a> {
...
-            } else {
-                Ok((dml::FieldType::Scalar(scalar_type, type_alias, None), vec![]))
-            }
-        } else if self.db.ast().find_model(type_name).is_some() {
-            Ok((dml::FieldType::Relation(dml::RelationInfo::new(type_name)), vec![]))
-        } else if self.db.get_enum(type_name).is_some() {
-            Ok((dml::FieldType::Enum(type_name.clone()), vec![]))
-        } else {
-            self.resolve_custom_type(ast_field, checked_types)
+fn lift_scalar_field_type(
```
Comment (tomhoule) [inline libs/datamodel/core/src/transform/ast_to_dml/lift.rs:183]: This is mostly just moved in its own function, one 2-3 levels of indentation lower. No significant change to review.

File: libs/datamodel/core/src/transform/ast_to_dml/lift.rs [hunk 2]
```diff
@@ -32,101 +30,120 @@ impl<'a> LiftAstToDml<'a> {
...
+            field.documentation = ast_field.documentation.clone().map(|comment| comment.text);
+            let mut field = dml::Field::RelationField(field);
+
+            if let Err(mut errs) = self.attributes.field.validate_and_apply(ast_field, &mut field) {
+                diagnostics.append(&mut errs);
+            }
+
+            field_ids_for_sorting.insert(&ast_field.name.name, field_id);
+            model.add_field(field)
         }
```
Comment (tomhoule) [inline libs/datamodel/core/src/transform/ast_to_dml/lift.rs:107]: Seeking feedback on these two loops for scalar and relation fields. These could easily live in methods with lots of parameters. I don't know what is more readable. I would assume the loop body will tend to get _shorter_ with time.

File: libs/datamodel/core/src/transform/ast_to_dml/lift.rs [hunk 3]
```diff
@@ -1,29 +1,27 @@
...
 use regex::Regex;
-use std::str::FromStr;
+use std::collections::{HashMap, HashSet};
 
 /// Helper for lifting a datamodel.
 ///
 /// When lifting, the AST is converted to the real datamodel, and additional
 /// semantics are attached.
 pub struct LiftAstToDml<'a> {
     attributes: AllAttributes,
```
Comment (pimeys) [inline libs/datamodel/core/src/transform/ast_to_dml/lift.rs:24]: This one next?

File: libs/datamodel/core/src/transform/ast_to_dml/lift.rs [hunk 4]
```diff
@@ -1,29 +1,27 @@
...
-    source: Option<&'a configuration::Datasource>,
```
Comment (pimeys) [inline libs/datamodel/core/src/transform/ast_to_dml/lift.rs:26]: 👍🏾

File: libs/datamodel/core/src/transform/ast_to_dml/lift.rs [hunk 5]
```diff
@@ -212,204 +178,181 @@ impl<'a> LiftAstToDml<'a> {
...
-            return Err(DatamodelError::new_validation_error(
-                &format!(
-                    "Recursive type definitions are not allowed. Recursive path was: {} -> {}.",
-                    checked_types.join(" -> "),
-                    type_name
-                ),
-                field_type.span,
-            ));
-        }
+fn lift_native_type(
```
Comment (pimeys) [inline libs/datamodel/core/src/transform/ast_to_dml/lift.rs:222]: This one next?

File: libs/datamodel/core/src/transform/ast_to_dml/lift.rs (hunk 1 above)
Comment (pimeys) [reply libs/datamodel/core/src/transform/ast_to_dml/lift.rs:183]: Five parameters. Would it still make sense to have this in the struct impl?

File: libs/datamodel/core/src/transform/ast_to_dml/lift.rs [hunk 6]
```diff
@@ -32,101 +30,120 @@ impl<'a> LiftAstToDml<'a> {
...
+    pub(crate) fn new(preview_features: &HashSet<PreviewFeature>, db: &'a ParserDatabase<'a>) -> LiftAstToDml<'a> {
         LiftAstToDml {
             attributes: AllAttributes::new(preview_features),
-            source,
             db,
         }
     }
 
-    pub fn lift(&self) -> Result<dml::Datamodel, Diagnostics> {
+    pub fn lift(&self, diagnostics: &mut Diagnostics) -> dml::Datamodel {
```
Comment (pimeys) [inline libs/datamodel/core/src/transform/ast_to_dml/lift.rs:40]: What about having the `&mut Diagnostics` in the constructor, then calling this with `self` by-value? Would cut one parameter out from the methods/functions...

File: libs/datamodel/core/src/transform/ast_to_dml/lift.rs [hunk 7]
```diff
@@ -32,101 +30,120 @@ impl<'a> LiftAstToDml<'a> {
...
-    fn lift_model(&self, ast_model: &ast::Model) -> Result<dml::Model, Diagnostics> {
+    fn lift_model(&self, model_id: ast::TopId, ast_model: &ast::Model, diagnostics: &mut Diagnostics) -> dml::Model {
         let mut model = dml::Model::new(ast_model.name.name.clone(), None);
         model.documentation = ast_model.documentation.clone().map(|comment| comment.text);
 
-        let mut errors = Diagnostics::new();
+        let active_connector = self.db.active_connector();
+        let mut field_ids_for_sorting: HashMap<&str, ast::FieldId> = HashMap::with_capacity(ast_model.fields.len());
+
+        for (field_id, scalar_field_type) in self.db.iter_model_scalar_fields(model_id) {
```
Comment (pimeys) [inline libs/datamodel/core/src/transform/ast_to_dml/lift.rs:64]: yiss.

File: libs/datamodel/core/src/transform/ast_to_dml/lift.rs [hunk 8]
```diff
@@ -32,101 +30,120 @@ impl<'a> LiftAstToDml<'a> {
...
+            if let Err(mut errs) = self.attributes.field.validate_and_apply(&attributes, &mut field) {
+                diagnostics.append(&mut errs);
             }
+            field_ids_for_sorting.insert(&ast_field.name.name, field_id);
+            model.add_field(field);
         }
 
-        if let Err(mut err) = self.attributes.model.validate_and_apply(ast_model, &mut model) {
-            errors.append(&mut err);
+        for (field_id, relation_field) in self.db.iter_model_relation_fields(model_id) {
```
Comment (pimeys) [inline libs/datamodel/core/src/transform/ast_to_dml/lift.rs:85]: lovely

File: libs/datamodel/core/src/transform/ast_to_dml/lift.rs (hunk 2 above)
Comment (pimeys) [reply libs/datamodel/core/src/transform/ast_to_dml/lift.rs:107]: No, just keep them like this.

File: libs/datamodel/core/src/transform/ast_to_dml/lift.rs [hunk 9]
```diff
@@ -32,101 +30,120 @@ impl<'a> LiftAstToDml<'a> {
...
+                diagnostics.append(&mut errs);
+            }
+
+            field_ids_for_sorting.insert(&ast_field.name.name, field_id);
+            model.add_field(field)
         }
 
-        if errors.has_errors() {
-            return Err(errors);
+        model.fields.sort_by_key(|f| field_ids_for_sorting.get(f.name()));
```
Comment (pimeys) [inline libs/datamodel/core/src/transform/ast_to_dml/lift.rs:109]: nice

File: libs/datamodel/core/src/transform/ast_to_dml/validation_pipeline.rs [hunk 10]
```diff
@@ -52,22 +52,18 @@ impl<'a, 'b> ValidationPipeline<'a> {
...
-                // Cannot continue on lifter error.
-                diagnostics.append(&mut err);
-                return Err(diagnostics);
-            }
-            Ok(schema) => schema,
-        };
+        let mut schema = lifter.lift(&mut diagnostics);
+
+        // Cannot continue on lifter error.
+        diagnostics.to_result()?;
```
Comment (pimeys) [inline libs/datamodel/core/src/transform/ast_to_dml/validation_pipeline.rs:66]: 👍🏾

**Review by pimeys — APPROVED** [review]
Yeah. Go.

File: libs/datamodel/core/src/transform/ast_to_dml/lift.rs (hunk 3 above)
Comment (tomhoule) [reply libs/datamodel/core/src/transform/ast_to_dml/lift.rs:24]: Yes, it's next.

File: libs/datamodel/core/src/transform/ast_to_dml/lift.rs (hunk 5 above)
Comment (tomhoule) [reply libs/datamodel/core/src/transform/ast_to_dml/lift.rs:222]: Hehe I'm a bit afraid to touch this one, but it doesn't look that scary anymore on its own. You can have a shot if you want. We'll get to it at some point.

File: libs/datamodel/core/src/transform/ast_to_dml/lift.rs (hunk 6 above)
Comment (tomhoule) [reply libs/datamodel/core/src/transform/ast_to_dml/lift.rs:40]: That's one path we can go. The other would be having `lift()` as a free function. Wdyt? I'm not sure.

Comment (pimeys) [reply libs/datamodel/core/src/transform/ast_to_dml/lift.rs:40]: Some of the functions have way too many params. Clippy says 7 or 8 is the maximum, but 5 already looks like a lot to my eyes.

Comment (tomhoule) [reply libs/datamodel/core/src/transform/ast_to_dml/lift.rs:40]: We'll go with that :+1: Commit incoming.

**Pull request merged**